### PR TITLE
Improve performance of keepalive rescheduling

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

If not all handlers were done, the keep alive would get rescheduled every second until the current time was greater than self._keepalive_time + self._keepalive_timeout

Instead we will reschedule the timer for the expected keep alive close time if the timer handle fires too early.

## Are there changes in behavior for the user?

no

## Is it a substantial burden for the maintainers to support this?
no

related issue #8613

Before we would see 1 timer handle created every second for 60s per connection. After we get 1 timer handle total for when the keepalive timeout is due.

before
before

after
~~after~~

after_take_2

### CODE REVIEW

**Comment by bdraco** [comment]
Before we would see 1 timer handle created every second for 60s per connection.  After we get 1 timer handle total.

**Comment by bdraco** [comment]
I went to do some archeology before I mark this ready as the previous code didn't make much sense to me.

**Comment by bdraco** [comment]
It looks like the 1s delay was added in https://github.com/aio-libs/aiohttp/commit/ad2366e31c0b05905082bd7b22340eb186128be0

Previously it was `next` but it looks like that caused a problem for websocket but thats since been removed.  It also looks like it caused a problem where it might fire too often, but thats not a problem with this implementation because we use `max(next, 1)`

File: CHANGES/8662.misc.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Reduce frequency of HTTP keep-alive checks -- by :user:`bdraco`.
```
Comment (bdraco) [inline CHANGES/8662.misc.rst]:
```suggestion
Improved performance of HTTP keep-alive checks -- by :user:`bdraco`.
```

File: aiohttp/web_protocol.py [hunk 2]
```diff
@@ -449,17 +449,19 @@ def _process_keepalive(self) -> None:
...
-                self.force_close()
-                return
+        if self._waiter and now > next:
+            self.force_close()
+            return
 
         # not all request handlers are done,
-        # reschedule itself to next second
-        self._keepalive_handle = self._loop.call_later(
-            self.KEEPALIVE_RESCHEDULE_DELAY,
```
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py:462]: Ironically, this was introduced to fix a CPU spike: #2701.

I wonder if we can improve this further and avoid the 1 second reschedules entirely? From what I can tell, the 1 second rechecks will be happening while a request is being processed (i.e. we are not waiting for a new message, so there is no _waiter).

Can't we just reschedule once the request handler has finished processing? It appears to me the waiter is only set in start(), and there's no way _keepalive_time won't get updated just before creating another waiter:
https://github.com/aio-libs/aiohttp/blob/4d604eafefbee22ad35886c3bd12e4c3048cc45b/aiohttp/web_protocol.py#L612

Therefore, it seems to me that if we reach this point and `now > next`, then there's no point in rescheduling here, we can just exit and let it schedule again after updating the timeout.

Comment (bdraco) [reply aiohttp/web_protocol.py:462]: I'll adjust this a bit to reschedule if it fires too early and avoid the 1s reschedules.

File: aiohttp/web_protocol.py [hunk 3]
```diff
@@ -445,23 +443,21 @@ def log_exception(self, *args: Any, **kw: Any) -> None:
...
+        self._keepalive_handle = None
         if self._force_close or not self._keepalive:
             return
 
-        next = self._keepalive_time + self._keepalive_timeout
+        loop = self._loop
+        now = loop.time()
+        close_time = self._next_keepalive_close_time
+        if now <= close_time:
+            # Keep alive fired close fired too early, reschedule
```
Comment (bdraco) [inline aiohttp/web_protocol.py]:
```suggestion
            # Keep alive close check fired too early, reschedule
```

File: CHANGES/8662.misc.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Improved performance of HTTP keep-alive checks -- by :user:`bdraco`.
+
+Previously, when processing a request for a keep-alive connection, the keep-alive check would happen every second; the check now occurs when the keep-alive is expected to timeout or 1s later, whichever is later.
```
Comment (bdraco) [inline CHANGES/8662.misc.rst]:
```suggestion
Previously, when processing a request for a keep-alive connection, the keep-alive check would happen every second; the check is now rescheduled if it fires too early instead.
```

**Review by Dreamsorcerer — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks. I like the new version much better.

**Pull request merged**